File: Ejercicio 12/EJercicio 12.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <iomanip>

using namespace std;

// Función para implementar Bucket Sort
void bucketSort(float arr[], int n) {
    vector<float> buckets[n];
    
    for (int i = 0; i < n; i++) {
        int bucketIndex = n * arr[i];
        buckets[bucketIndex].push_back(arr[i]);
    }
    
    for (int i = 0; i < n; i++) {
        sort(buckets[i].begin(), buckets[i].end());
    }
    
    int index = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < buckets[i].size(); j++) {
            arr[index++] = buckets[i][j];
        }
    }
}

float generarAleatorio() {
    return static_cast<float>(rand() % 1000) / 1000.0f;
}

void imprimirArray(float arr[], int n) {
    for (int i = 0; i < n; i++) {
        cout << fixed << setprecision(3) << arr[i] << " ";
    }
    cout << endl;
}

int main() {
    srand(time(0));
    
    const int TAMANO = 15;
    float arr[TAMANO];
    
    // Generar 15 números aleatorios entre 0 y 1
    cout << "Generando 15 numeros aleatorios entre 0 y 1" << endl;
    for (int i = 0; i < TAMANO; i++) {
        arr[i] = generarAleatorio();
    }
    
    // Mostrar array original
    cout << "\nArray original:" << endl;
    imprimirArray(arr, TAMANO);
    
    // Aplicar Bucket Sort
    bucketSort(arr, TAMANO);
    
    // Mostrar array ordenado
    cout << "\nArray ordenado con Bucket Sort:" << endl;
    imprimirArray(arr, TAMANO);
    
    return 0;
}

File: Ejercicio 12/EJercicio 12.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    
    static void BucketSort(float[] arr, int n)
    {
        List<float>[] buckets = new List<float>[n];
        
        for (int i = 0; i < n; i++)
        {
            buckets[i] = new List<float>();
        }
        
        for (int i = 0; i < n; i++)
        {
            int bucketIndex = (int)(n * arr[i]);
            buckets[bucketIndex].Add(arr[i]);
        }
        
        for (int i = 0; i < n; i++)
        {
            buckets[i].Sort();
        }
        
        int index = 0;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < buckets[i].Count; j++)
            {
                arr[index++] = buckets[i][j];
            }
        }
    }

    static float GenerarAleatorio()
    {
        Random rand = new Random();
        return (float)rand.Next(1000) / 1000.0f;
    }

    static void ImprimirArray(float[] arr, int n)
    {
        for (int i = 0; i < n; i++)
        {
            Console.Write(arr[i].ToString("F3") + " ");
        }
        Console.WriteLine();
    }

    static void Main()
    {
        const int TAMANO = 15;
        float[] arr = new float[TAMANO];
        
        // Generar 15 números aleatorios entre 0 y 1
        Console.WriteLine("Generando 15 numeros aleatorios entre 0 y 1");
        for (int i = 0; i < TAMANO; i++)
        {
            arr[i] = GenerarAleatorio();
        }
        
        // Mostrar array original
        Console.WriteLine("\nArray original:");
        ImprimirArray(arr, TAMANO);
        
        // Aplicar Bucket Sort
        BucketSort(arr, TAMANO);
        
        // Mostrar array ordenado
        Console.WriteLine("\nArray ordenado con Bucket Sort:");
        ImprimirArray(arr, TAMANO);
    }
}